### PR TITLE
[Fix #941] Add support for ignoring certain variables in RSpec/VariableName

### LINKED ISSUE #941: RSpec/VariableName: Would appreciate an 'ignore' configuration option
**Feature Request**
A configuration option to specify certain variables or patterns to ignore

**Example**
```yaml
RSpec/VariableName:
  Enabled: true
  EnforcedStyle: "snake_case"
  IgnoreVariables:
    - fooBar
    - /fizzBuzz/
```
**Context**
We use an RSpec plugin called [RSwag](https://github.com/rswag/rswag) that allows the creation of specs & swagger documentation from a single DSL. One of the features is that documented parameters can be populated in tests via `let` statements

eg.
```ruby
describe 'Blogs API - Individual Blog Endpoint' do
  path '/blogs/{id}' do
	parameter(name: :id, in: :path, required: true, type: :string)
    let(:id) { 'foo' }
	# test code ensues
  end
end
```

The issue arises when providing an array as a query parameter
```ruby
describe 'Blogs API - Blog Collection Endpoint' do
  path '/blogs' do
    # making a request to /blogs?id[]=foo&id[]=bar&id[]=baz
    parameter(
      name: :'id[]',
      in: :query,
      required: false,
      type: :array,
      items: { type: :string },
      collectionFormat: :multi,
    )
    let(:'id[]') { ['foo', 'bar', 'baz'] } # sets off the cop
    # test code ensues
  end
end 
```

It's way too much of a corner case to modify the cop to ignore that specific pattern, but I think a generic option to ignore certain things by name or regex is appropriate.

### PR DESCRIPTION
**Pull request by jtannas**
Fixes #941

This PR adds support for ignoring naming conventions on certain RSpec variables via configuration options. Options for `IgnoredVariables` and `IgnoredPatterns` are provided, and are empty by default. This will be helpful for strange cases and for codebases that are iteratively refactoring to apply the rule.

My original thought was to have them combined in a single `Ignore` option which would accept both static values and regex patterns, but the extra complexity of parsing the configuration led me to split it in two.

Edit: `IgnoredVariables` has been removed - see comments below

---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Updated documentation.
* [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have modified an existing cop's configuration options:

* [x] Set `VersionChanged` in `default/config.yml` to the next major version.

### CODE REVIEW

File: lib/rubocop/cop/rspec/variable_name.rb [hunk 1]
```diff
@@ -14,6 +17,14 @@ module RSpec
       #   let(:user_name) { 'Adam' }
       #   subject(:user_name) { 'Adam' }
       #
+      #   # okay with `userFood` listed in `IgnoredVariables`
```
Comment (pirj) [inline lib/rubocop/cop/rspec/variable_name.rb:20]: Typically we describe it in a separate `@example`:
```ruby
      # @example `Prefixes` configuration
      #
      #   # .rubocop.yml
      #   # RSpec/ContextWording:
      #   #   Prefixes:
      #   #     - when
      #   #     - with
      #   #     - without
      #   #     - if
      #   #     - unless
      #   #     - for
      #
      # @example
      #   # bad
      #   context 'the display name not present' do
      #     # ...
      #   end
      #
      #   # good
      #   context 'when the display name is not present' do
      #     # ...
      #   end
```

do you mind adding two examples to demonstrate how we expect the configuration of those options?
Otherwise, it may create a false impression that `IgnoredVariables`/`IgnoredPatterns` only relate to `EnforcedStyle: snake_case`.

File: lib/rubocop/cop/rspec/variable_name.rb [hunk 2]
```diff
@@ -24,13 +35,16 @@ module RSpec
       #   subject(:userName) { 'Adam' }
       class VariableName < Base
         include ConfigurableNaming
+        include IgnoredPattern
```
Comment (pirj) [inline lib/rubocop/cop/rspec/variable_name.rb:38]: 👍

File: lib/rubocop/cop/rspec/variable_name.rb [hunk 3]
```diff
@@ -41,6 +55,10 @@ def on_send(node)
         def message(style)
           format(MSG, style: style)
         end
+
+        def matches_ignored_variable?(value)
```
Comment (pirj) [inline lib/rubocop/cop/rspec/variable_name.rb:59]: Maybe `value` -> `variable_name`?

File: spec/rubocop/cop/rspec/variable_name_spec.rb [hunk 4]
```diff
@@ -68,6 +72,18 @@
           let("user#{name}") { 'Adam' }
         RUBY
       end
+
+      it 'does not register offense for configured ignoredVariable' do
+        expect_no_offenses(<<~'RUBY')
+          let('ignoredVariable') { 'Adam' }
+        RUBY
+      end
```
Comment (pirj) [inline spec/rubocop/cop/rspec/variable_name_spec.rb:80]: Please add those examples separate from `context 'when configured for `snake_case`' do`, along with config setup, and preferably each having a balancing example where the same variable name is flagged.

**Comment by pirj** [comment]
Do you think we really need `IgnoredVariables` at all given that it's possible to add `^var_ignored$` to `IgnoredPatterns`?
I have no preference here, but I can't figure out a good use case for it. What did you have in mind, where would it come handy in your practice?

File: CHANGELOG.md [hunk 5]
```diff
@@ -8,6 +8,7 @@
 * Relax `RSpec/VariableDefinition` cop so interpolated and multiline strings are accepted even when configured to enforce the `symbol` style. ([@bquorning][])
 * Fix `RSpec/EmptyExampleGroup` to flag example groups with examples in invalid scopes. ([@mlarraz][])
 * Fix `RSpec/EmptyExampleGroup` to ignore examples groups with examples defined inside iterators. ([@pirj][])
+* Add `IgnoreVariables` and `IgnorePatterns` configuration options to`RSpec/VariableName. ([@jtannas][])
```
Comment (pirj) [inline CHANGELOG.md:11]: "to`RSpec/VariableName" - backtick is missing, space is missing.

**Comment by jtannas** [comment]
> Do you think we really need `IgnoredVariables` at all given that it's possible to add `^var_ignored$` to `IgnoredPatterns`?
> I have no preference here, but I can't figure out a good use case for it. What did you have in mind, where would it come handy in your practice?

I suppose we don't need both. My original thought was that the `IgnoredVariables` is more user-friendly for simple cases than using regex patterns. In hindsight though, having redundant configuration options is probably even less user friendly.
I'll drop the `IgnoredVariables` since `IgnoredPatterns` is more flexible/powerful/already written.

File: spec/rubocop/cop/rspec/variable_name_spec.rb [hunk 6]
```diff
@@ -188,4 +188,33 @@
...
   end
+
+  context 'when configured to ignore certain patterns' do
+    context 'when the variable does not match any of the patterns' do
+      let(:cop_config) do
+        { 'EnforcedStyle' => 'snake_case',
+          'IgnoredPatterns' => ['^userFood$', '^userPet$'] }
+      end
+
+      it 'registers an offense' do
```
Comment (bquorning) [inline spec/rubocop/cop/rspec/variable_name_spec.rb:199]: Sorry for nitpicking, but I’ll suggest removing the innermost `context`s, and use the same `cop_config` for these two examples. E.g.:
```rb
context 'when configured to ignore certain patterhs' do
  let(:cop_config) do
    # …
  end

  it 'registers an offense when not matching any ignored pattern' do
    # …
  end

  it 'registers an offense when matching at least one ignored pattern' do
    # …
  end
end
```

What do you think?

File: lib/rubocop/cop/rspec/variable_name.rb [hunk 7]
```diff
@@ -22,15 +25,31 @@ module RSpec
...
+      #   # rubocop.yml
+      #   # RSpec/VariableName:
+      #   #   EnforcedStyle: snake_case
+      #   #   IgnoredPatterns:
+      #   #     - ^userFood$
+      #
+      # @example
+      #   # okay because it matches the`^userFood$` regex in `IgnoredPatterns`
+      #   let(:userFood) { 'spaghetti' }
+      #   subject(:userFood) { 'spaghetti' }
```
[the remark this answers is not in the document]
Comment (jtannas) [reply lib/rubocop/cop/rspec/variable_name.rb:40]: sounds good - I was using the pattern from the existing examples, so I'll go ahead and make the same fixes to them while I'm in there

**Comment by bquorning** [comment]
Could you squash the 3 commits into one, please? Then we’re ready to merge 👍
